be/src/runtime/sender_queue.cpp

```diff
@@ -0,0 +1,718 @@
+                _chunk_queue.emplace_back(std::move(item));
+            }
+            bool has_new_chunks = _chunk_queue.size() > original_size;
+            if (has_new_chunks && done != nullptr && _recvr->exceeds_limit(total_chunk_bytes)) {
+                _chunk_queue.back().closure = *done;
+                _chunk_queue.back().queue_enter_time = MonotonicNanos();
+                COUNTER_UPDATE(_recvr->_closure_block_counter, 1);
+                *done = nullptr;
+            }
+            _recvr->_num_buffered_bytes += total_chunk_bytes;
+            _data_arrival_cv.notify_one();
+        } else {
+            auto& chunk_queues = _buffered_chunk_queues[be_number];
+
+            if (!chunks.empty() && done != nullptr && _recvr->exceeds_limit(total_chunk_bytes)) {
+                chunks.back().closure = *done;
+                chunks.back().queue_enter_time = MonotonicNanos();
+                COUNTER_UPDATE(_recvr->_closure_block_counter, 1);
+                *done = nullptr;
+            }
+
+            // The queue in chunk_queues cannot be changed, so it must be
+            // assigned to chunk_queues after local_chunk_queue is initialized
+            // Otherwise, other threads may see the intermediate state because
+            // the initialization of local_chunk_queue is beyond mutex
+            chunk_queues[sequence] = std::move(chunks);
+
+            phmap::flat_hash_map<int64_t, ChunkQueue>::iterator it;
+            int64_t& max_processed_sequence = _max_processed_sequences[be_number];
+
+            // max_processed_sequence + 1 means the first unprocessed sequence
+            while ((it = chunk_queues.find(max_processed_sequence + 1)) != chunk_queues.end()) {
+                ChunkQueue& unprocessed_chunk_queue = (*it).second;
+
+                // Now, all the packets with sequance <= unprocessed_sequence have been received
+                // so chunks of unprocessed_sequence can be flushed to ready queue
+                for (auto& item : unprocessed_chunk_queue) {
+                    _chunk_queue.emplace_back(std::move(item));
+                }
+
+                chunk_queues.erase(it);
+                ++max_processed_sequence;
+            }
+
+            _recvr->_num_buffered_bytes += total_chunk_bytes;
+        }
+    }
+    return Status::OK();
+}
+
+void DataStreamRecvr::NonPipelineSenderQueue::decrement_senders(int be_number) {
+    std::lock_guard<Mutex> l(_lock);
+    if (_sender_eos_set.end() != _sender_eos_set.find(be_number)) {
+        return;
+    }
+    _sender_eos_set.insert(be_number);
+    DCHECK_GT(_num_remaining_senders, 0);
+    _num_remaining_senders--;
+    VLOG_FILE << "decremented senders: fragment_instance_id=" << print_id(_recvr->fragment_instance_id())
+              << " node_id=" << _recvr->dest_node_id() << " #senders=" << _num_remaining_senders
+              << " be_number=" << be_number;
+    if (_num_remaining_senders == 0) {
+        _data_arrival_cv.notify_one();
```

Comment:
Why notify_one not notify_all

Reply:
> Why notify_one not notify_all

this is the original logic, I think notify_all makes more sense, I have fixed it.